test/test-mc-fle2-payload-iev.c

```diff
@@ -269,6 +269,60 @@ test_FLE2IndexedEqualityEncryptedValue_decrypt (_mongocrypt_tester_t *tester)
       mongocrypt_status_destroy (status);
    }
 
+   /* Test that parse and write work correctly. */
```

Comment:
This tests serializes a parsed payload. Is `mc_FLE2IndexedEncryptedValue_write` expected to serialize a payload from parts?

I expect callers of `mc_FLE2IndexedEncryptedValue_write` will pass the necessary parameters (tokens, ClientEncryptedValue, counter, etc.), rather than parsing and decrypting an existing payload.

Consider changing `mc_FLE2IndexedEncryptedValue_write` to accept the necessary parameters, instead of using `mc_FLE2IndexedEncryptedValue_t`:

```c
// mc_FLE2IndexedEqualityEncryptedValue_write serializes and encrypts an FLE2IndexedEqualityEncryptedValue. 
bool
mc_FLE2IndexedEqualityEncryptedValue_write (_mongocrypt_crypto_t *crypto,
                                            mc_FLE2IndexedEqualityEncryptedValueTokens *index_tokens,
                                            bson_type_t original_bson_type,
                                            _mongocrypt_buffer_t *S_KeyId,
                                            _mongocrypt_buffer_t ClientEncryptedValue,
                                            mc_ServerDataEncryptionLevel1Token_t *token,
                                            _mongocrypt_buffer_t *buf,
                                            mongocrypt_status_t *status);
```

That may more closely model the server [`FLE2IndexedEqualityEncryptedValue` constructor](https://github.com/mongodb/mongo/blob/ac877eecf0c5138b7f0a057ba86c92d9aeb284e7/src/mongo/crypto/fle_crypto.h#L746-L753).

Reply:
The callers of mc_FLE2IndexedEncryptedValue_write will likely pass in a InsertUpdate payload and the counts value. See [this patch of code](https://github.com/mongodb/mongo/blob/master/src/mongo/crypto/fle_crypto.cpp#L1528-L1544) for reference. 

In that patch, I was going to keep line 1528, serialize the data out to a binary string, and parse it as a mc_FLE2IndexedEqualityEncryptedValue_t, then call write. In effect do two conversions, which will greatly reduce the amount of code duplication in the future when we do [SERVER-71338](https://jira.mongodb.org/browse/SERVER-71338).

However, your comment makes a lot of sense in that we should be constructing the mc_FLE2IndexedEqualityEncryptedValue_t object in the function. A lot of the code is going to be duplicated and I was hoping to avoid that, but I do not think that it is worth the perf hit that I think my way would cause.

Reply:
Refactored the code. I think it makes more sense.